### PR TITLE
Buffers refactoring

### PR DESCRIPTION
**Pull request by prokopyl**
The original goal of this PR was to clean up and extract all of buffer-manipulation logic that was surrounding the plugin's `process` function. In the process of doing that, this PR also does the following:

* Make the `SharedBuffer<T>` implementation more generic: this removes specific implementations for `f32` and `f64` sample types and event types, and replaces them with generic implementations that work for any time (requiring `Default` when necessary).
* Remove all of the constant mask computation and caching: this was a missed leftover of the Clack migration: the constant mask is a CLAP-specific feature that only makes sense when independent audio buffers are grouped together as channels at the edge of Plugin processing cycles. Since we store our `is_constant` flags on each buffer and Clack already uses these to recompute the constant mask for each process cycle, the cached value was sitting unused and can now be safely removed.
* Change the parameter and note buffers to use custom types instead of CLAP events. This allows to better control out-of-band parameters (such as port IDs) and also remove potentially invalid events to transit over those buffers. These types also now have specific methods to convert to and from CLAP events, allowing more separation between it and the processing logic.

### CODE REVIEW

File: core/src/plugin/buffer.rs [hunk 1]
```diff
@@ -169,254 +142,60 @@ pub enum RawAudioChannelBuffers {
...
-                        for x in buf.iter() {
-                            if *x != 0.0 {
-                                return false;
-                            }
-                        }
-                    } else {
-                        return false;
                     }
                 }
             }
```
Comment (BillyDM) [inline core/src/plugin/buffer.rs]: Why did you remove the ability for a quick silence check using the constant flag?

The purpose of the `use_slow_check` argument is so the plugin can decide if it only wants to use the fast check with the constant flag, or if it wants to do a thorough check.

File: core/src/plugin/buffer.rs [hunk 2]
```diff
@@ -1,91 +1,73 @@
...
-            DebugBufferType::Audio32 => write!(f, "f"),
-            DebugBufferType::Audio64 => write!(f, "d"),
-            DebugBufferType::IntermediaryAudio32 => write!(f, "if"),
-            DebugBufferType::ParamAutomation => write!(f, "pa"),
-            DebugBufferType::NoteBuffer => write!(f, "n"),
+            DebugBufferType::Audio32 => f.write_str("f32"),
+            DebugBufferType::Audio64 => f.write_str("f64"),
+            DebugBufferType::IntermediaryAudio32 => f.write_str("intermediary_f32"),
+            DebugBufferType::Event => f.write_str("event"),
+            DebugBufferType::Note => f.write_str("note"),
```
Comment (BillyDM) [inline core/src/plugin/buffer.rs]: The purpose of having custom debugging here is so we can compactly debug the entire schedule. I want to aim for each task in the schedule fitting in a single line.

Change this do something like:
```rust
          DebugBufferType::Audio32 => f.write_str("fl"),
          DebugBufferType::Audio64 => f.write_str("db"),
          DebugBufferType::IntermediaryAudio32 => f.write_str("ifl"),
          DebugBufferType::Event => f.write_str("ev"),
          DebugBufferType::Note => f.write_str("nt"),
```

File: core/src/plugin/process_info.rs [hunk 3]
```diff
@@ -50,18 +50,9 @@ pub struct ProcBuffers {
...
 impl ProcBuffers {
-    pub fn audio_inputs_silent(&self, use_slow_check: bool, frames: usize) -> bool {
+    pub fn audio_inputs_silent(&self, frames: usize) -> bool {
         for buf in self.audio_in.iter() {
-            if !buf.is_silent(use_slow_check, frames) {
-                return false;
-            }
-        }
-        true
-    }
```
Comment (BillyDM) [inline core/src/plugin/process_info.rs]: Again, I want plugins to be able to use either a quick check with the constant masks, or a thorough check.

File: Cargo.toml [hunk 4]
```diff
@@ -26,19 +26,18 @@ cpal-asio = ["cpal/asio"]
...
 meadowlark-core-types = "0.3"
 audio-graph = { git = "https://github.com/MeadowlarkDAW/audio-graph", branch = "main" }
-clack-host = { git = "https://github.com/prokopyl/clack", rev = "d2c0f2e43b56e7037cb96729cebfb40fbc868c51" }
-clack-extensions = { git = "https://github.com/prokopyl/clack", features = ["clack-host", "audio-ports", "log", "note-ports", "params", "state", "thread-check"], rev = "d2c0f2e43b56e7037cb96729cebfb40fbc868c51" }
+clack-host = { git = "https://github.com/prokopyl/clack", rev = "782bbfce72f6b3505dc70807bbebadcb3dc5ac1b" }
+clack-extensions = { git = "https://github.com/prokopyl/clack", features = ["clack-host", "audio-ports", "log", "note-ports", "params", "state", "thread-check"], rev = "782bbfce72f6b3505dc70807bbebadcb3dc5ac1b" }
 dirs = { version = "4.0", optional = true }
 basedrop = "0.1"
-smallvec = "1.7"
+smallvec = { version = "1.9.0", features = ["const_generics", "union"] }
```
Comment (BillyDM) [inline Cargo.toml:33]: What's the advantage of using const_generics and unions here?

**Review by BillyDM — CHANGES_REQUESTED** [review]

File: core/src/plugin/buffer.rs (hunk 1 above)
Comment (prokopyl) [reply core/src/plugin/buffer.rs]: Ah, I didn't realize this was an API to be exposed to plugins, I only saw `use_slow_check` set to true in all usages so I just went ahead and removed all the dead code.

I'll re-add the fast implementation today. :slightly_smiling_face:

File: Cargo.toml (hunk 4 above)
Comment (prokopyl) [reply Cargo.toml:33]: [`const_generics`](https://docs.rs/smallvec/1.9.0/smallvec/#const_generics) allows to use buffers of any size, and probably helps a little bit with compile times since it's a single impl vs. a large amount of macro-generated impls for each array size (to a point). 

[`union`](https://docs.rs/smallvec/1.9.0/smallvec/#union) is the most useful of the two since it makes `SmallVec`s smaller by 1 word (8 bytes on 64bit) by removing the internal enum tag, which I found stacks up quite fast since we're using `SmallVec`s of `SmallVec`s to store buffer pointers.
